File: 프로그래머스/3/42895. N으로 표현/N으로 표현.py
def solution(N, number):
    if N == number:
        return 1
    
    answer = -1
    dp = [set() for _ in range(8)]
    for i in range(8):
        dp[i].add(int(str(N)*(i+1)))
    
    for i in range(1, 8):
        for j in range(i):
            for op1 in dp[j]:
                for op2 in dp[i-j-1]:
                    dp[i].add(op1+op2)
                    dp[i].add(op1-op2)
                    dp[i].add(op1*op2)
                    if op2 != 0:
                        dp[i].add(op1//op2)
        if number in dp[i]:
                return i + 1
    
    return answer

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.js
function solution(N, number) {
    if (N===number) return 1;
    
    const dp = Array.from(new Array(9), ()=> new Set());
    
    dp.forEach((el, i) => {
        if (i !== 0){
            el.add(Number(String(N).repeat(i)));
        }
    })
    
    for (let i=1; i<=8; i++){
        for (let j=1; j<i; j++) {
            for (a of dp[j]) {
                for (b of dp[i-j]){
                    dp[i].add(a+b);
                    dp[i].add(a-b);
                    dp[i].add(a*b);
                    dp[i].add(Math.floor(a/b));
                }
            }
        }
        if (dp[i].has(number)) return i;
    }
    return -1;
    
}